Adding descriptions for each option

iamalisalehi opened the pull request:
Hi,

This patch  provides a feature to show descriptions for each options. The description for each option shows on the right side of the screen as the cursor moves on to it.
A function was also added that divides long description into multiple lines based on the terminal size.
An example was also provided.



Hope it is useful.

P.S. I couldn't think of a way to write a test for it so I didn't.

iamalisalehi commented:
Here's a gif of the example:
![descriptions](https://github.com/wong2/pick/assets/49132329/0c3ae8be-ed1e-4e64-abe6-8fade28a953e)

SnowzNZ commented:
I feel like it would be better to have descriptions and options stored as a dictionary rather than two lists.

gustavo-mag commented:
unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/AUX7SDUCK4CSA6RXM6LLD4LY2MRPPAVCNFSM6AAAAABFJWZTKGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMRTHA3TIMRWGY__;!!JVh_Qfuk4otXm3Mt0g!67AzWURm-WTfVyJpsuZ6U4uEqcwCB4llajiC0go_pxsS-Rk54x5KkK2GxpWOwPz4UXEQAqN5UFxxxyIcx71tsnPXNSO_Aakh$>



Internal Use Only

From: Snowz ***@***.***>
Sent: quarta-feira, 27 de março de 2024 17:09
To: wong2/pick ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [wong2/pick] Adding descriptions for each option (PR #114)


CAUTION: External email.

I feel like it would be better to have descriptions and options stored as a dictionary rather than two lists.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/wong2/pick/pull/114*issuecomment-2023874266__;Iw!!JVh_Qfuk4otXm3Mt0g!67AzWURm-WTfVyJpsuZ6U4uEqcwCB4llajiC0go_pxsS-Rk54x5KkK2GxpWOwPz4UXEQAqN5UFxxxyIcx71tsnPXNVZB2ZlH$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/AUX7SDUCK4CSA6RXM6LLD4LY2MRPPAVCNFSM6AAAAABFJWZTKGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMRTHA3TIMRWGY__;!!JVh_Qfuk4otXm3Mt0g!67AzWURm-WTfVyJpsuZ6U4uEqcwCB4llajiC0go_pxsS-Rk54x5KkK2GxpWOwPz4UXEQAqN5UFxxxyIcx71tsnPXNSO_Aakh$>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.******@***.***>>

iamalisalehi commented:
> I feel like it would be better to have descriptions and options stored as a dictionary rather than two lists.

That's a nice idea. I'll try it.

iamalisalehi commented:
So I just added a new commit which incorporates @SnowzNZ's idea with minimal changes to the way the code is written to still be backwards compatible (Although the descriptions will only work properly with python3.7+ because the dictionaries weren't ordered before that).

please read the commit notes too.

SnowzNZ reviewed (changes requested):
Remove `.idea` folder

wong2 commented:
In case you don't know, the item passed to `options` could be an `Option` instance, so I think it's better to add `description` as an attribute of `Option`. https://github.com/wong2/pick/blob/master/src/pick/__init__.py#L22

aisk commented on example/description.py:
```diff
@@ -0,0 +1,14 @@
+    "Haskell": ""
+}
+option, index = pick(options, title, indicator="=>", default_index=5)
+print(f"You choosed {option} at index {index}")
```
The ending new line is required. There is another file have the same issue.

iamalisalehi replied:
Thanks for the review, with the new commits it should be fixed. Also changed "choosed" to "chose" in all examples.

aisk commented:
Hi @iamalisalehi, if you want to continue the work, please merge the latest code from the master branch, which has fixed the CI, and we should ensure that it passes before merging.

iamalisalehi commented:
Hi @wong2 , thanks for the review.
Sorry I am not still very skilled in git and messed up commit messages a little bit so I'll explain what I did here:

1) Added an optional attribute `description` to the `Option` class.
2) Changed `parse_options` function to turn dictionaries in instances of `Option`. Now we don't need `self.descriptions` anymore.
3) Added a new example.
4) Updated `test_option` to include the new possibilities.
5) Changed the test for parsing dictionaries.

P.S. I didn't change `value` attribute, only gave it a default `None` value. Is this okay?

aisk commented on example/description_option.py:
```diff
@@ -0,0 +1,12 @@
+from pick import pick, Option
```
For this example, I'd like to reuse the existing `option.py` example.

aisk commented on src/pick/__init__.py:
```diff
@@ -48,6 +49,22 @@ def __post_init__(self) -> None:
+                    options.append(Option(option, description=description))
+
+            self.options = options
+
+        else:
+            pass
```
The `else` branch is unnecessary and should be removed. Additionally, I would like to apply the [early return principle](https://dev.to/shameel/-clean-coding-tip-early-return-principle-47kj) to make the code cleaner.

aisk commented on example/description_dict.py:
```diff
@@ -0,0 +1,14 @@
+    "PHP": "PHP is a general-purpose scripting language geared towards web development.",
+    "C++": "C++ is a high-level, general-purpose programming language created by Danish computer scientist Bjarne Stroustrup, first released in 1985 as an extension of the C programming language.",
+    "Erlang": "",
+    "Haskell": ""
```
```suggestion
    "Haskell": "",
```

aisk commented on src/pick/__init__.py:
```diff
@@ -129,9 +165,16 @@ def draw(self, screen: "curses._CursesWindow") -> None:
 
         for line in lines_to_draw:
-            screen.addnstr(y, x, line, max_x - 2)
+            screen.addnstr(y, x, line, max_x//2 - 2)
```
I think we should add a white space besides `//`, like other binary operators.

aisk commented on tests/test_pick.py:
```diff
@@ -48,10 +48,47 @@ def test_multi_select():
 
 def test_option():
-    options = [Option("option1", 101), Option("option2", 102), Option("option3", 103)]
+    options = [Option("option1", 101, "description1"), Option("option2", 102),
```
There is too many duplicated codes for this test case, maybe we can use some way to reduce them, like check the value in a for loop.

aisk reviewed (commented):
There are some comments for the PR, and the CI should be passed, I see it failed mainly because the type checker, you can run `mypy` on your local machine to fix them.

snake736 reviewed (commented):
Lecture

SnowzNZ commented:
> Lecture

?

aisk commented:
Actually, the current codes don't have the description on my machine. I guess it's been broken during the updated commit. This is the result of running `python example/description_dict.py`:

![image](https://github.com/wong2/pick/assets/699636/3fb59015-7216-4602-9f4c-2eaf4e0262b1)

aisk commented:
When picking a dict, the pick result is an `Option`. But in the current main branch, you pass a list of `str` of `Option`, and the pick result will be the original type.

So I think the pick result of a `dict` should be a two-element `tuple`, like when iterating over a `dict`.

iamalisalehi commented:
> Actually, the current codes don't have the expected behavior on my machine. I guess it's been broken during the updated commits. This is the result of running `python example/description_dict.py`:
> 
> ![image](https://private-user-images.githubusercontent.com/699636/319106390-3fb59015-7216-4602-9f4c-2eaf4e0262b1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIxNTI1NjYsIm5iZiI6MTcxMjE1MjI2NiwicGF0aCI6Ii82OTk2MzYvMzE5MTA2MzkwLTNmYjU5MDE1LTcyMTYtNDYwMi05ZjRjLTJlYWY0ZTAyNjJiMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQwM1QxMzUxMDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMWVmMGZkMzBhZDgwN2I2MmUyY2JiMjIwNTVjNDgyZWMzZTM2NDc5MWM1ZjZhODA0NzkwMWRjMTRiYTI3NWMzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.4xFTysVH_wUUsH9uhaHjkBTuEln8uiOyaNCvLaGsHeo)

Hi @aisk , I fixed the issues you raised, and checked the example again. It worked well on my computer (I am running python3.11.8). Can you please check it again?

iamalisalehi commented:
> When picking a dict, the pick result is an `Option`. But in the current main branch, you pass a list of `str` of `Option`, and the pick result will be the original type.
> 
> So I think the pick result of a `dict` should be a two-element `tuple`, like when iterating over a `dict`.

Hmm, I thought about this but I am not so sure that it would be the best way to do it. As I understood from the code before I worked on it, there is a simple way and a more complex way to use it. In the first versions that I wrote, adding descriptions was considered a part of the simple way of doing things but then at the suggestion of @wong2 , I changed it to be a part of the complex way of doing it (this also makes more sense to me).
But I feel like that returning a tuple, introduces a third way of doing things and needs another datatype to be considered and might introduce unnecessary complexity. But as you have probably seen, I don't have a lot of coding experience so I am not sure..

aisk commented:
> Hi @aisk , I fixed the issues you raised, and checked the example again. It worked well on my computer (I am running python3.11.8). Can you please check it again?

It's the same result, but I'm using Linux with Tilix as terminal emulator. Maybe this is OS dependent issue, I'll check the result in other OS tomorrow.

iamalisalehi commented:
> > Hi @aisk , I fixed the issues you raised, and checked the example again. It worked well on my computer (I am running python3.11.8). Can you please check it again?
> 
> It's the same result, but I'm using Linux with Tilix as terminal emulator. Maybe this is OS dependent issue, I'll check the result in other OS tomorrow.
> 
> Update: Kitty on Linux have no issue:

Does the `option.py` example work on Tilix or is it just the dict example?

P.S. I tested them on Konsole, Terminator and Pycharm's terminal and they work

aisk commented on src/pick/__init__.py:
```diff
@@ -129,9 +163,16 @@ def draw(self, screen: "curses._CursesWindow") -> None:
+            description_lines = self.get_description_lines(option.description, max_x // 2)
+
+            for i, line in enumerate(description_lines):
+                screen.addstr(i + 3, max_x // 2, line, 2 * max_x // 2 - 2)
```
```suggestion
                screen.addnstr(i + 3, max_x // 2, line, 2 * max_x // 2 - 2)
```

aisk reviewed (commented):
I resolved the problem; it's not caused by Tilix, but by a wider screen. With a wider screen, the last value passed to `addstr` in `draw` method will be larger than 256, which will draw the description text with the same color as the background, leading to invisibility.

Kitty works fine on my machine because the default font is larger than Tilix's, so the screen is not wide enoght to triger the bug.

I guess you want to use `addnstr` instead of `addstr` to specify the maximum line width, but you missed an `n`, which caused the bug.

aisk commented:
> > When picking a dict, the pick result is an `Option`. But in the current main branch, you pass a list of `str` of `Option`, and the pick result will be the original type.
> > So I think the pick result of a `dict` should be a two-element `tuple`, like when iterating over a `dict`.
> 
> Hmm, I thought about this but I am not so sure that it would be the best way to do it. As I understood from the code before I worked on it, there is a simple way and a more complex way to use it. In the first versions that I wrote, adding descriptions was considered a part of the simple way of doing things but then at the suggestion of @wong2 , I changed it to be a part of the complex way of doing it (this also makes more sense to me). But I feel like that returning a tuple, introduces a third way of doing things and needs another datatype to be considered and might introduce unnecessary complexity. But as you have probably seen, I don't have a lot of coding experience so I am not sure..

I have some slight concerns about the complexity too, so I think **for now**, maybe we should only support using `list[pick.Option]` as the input type for items with descriptions.

In this way, we can ensure that the `pick`'s result type matches the input type exactly. If you put a list of `str`, you'll get a `str`; if you put a list of `Option`, you'll get an `Option`.

iamalisalehi commented:
> I have some slight concerns about the complexity too, so I think **for now**, maybe we should only support using `list[pick.Option]` as the input type for items with descriptions.

So I should get rid of the `parse_options` function completely?

> In this way, we can ensure that the `pick`'s result type matches the input type exactly. If you put a list of `str`, you'll get a `str`; if you put a list of `Option`, you'll get an `Option`.

Question: In the `option` example now we see both `str` and `Option` inputs. Should this be acceptable? It outputs exactly the same thing and yet the input is not uniform.

aisk commented:
> So I should get rid of the `parse_options` function completely?

Yes, and thus we can revert specify the `self.options` to `Any`.
  
> Question: In the `option` example now we see both `str` and `Option` inputs. Should this be acceptable? It outputs exactly the same thing and yet the input is not uniform.

You reminded me, I think it's not accepted, because I think mypy or pyright will complain about this, I'm not using my development machine so I don't try it, but you can have a try. I think we should add a type checker to the examples on the GHA.

iamalisalehi commented:
> > So I should get rid of the `parse_options` function completely?
> 
> Yes, and thus we can revert specify the `self.options` to `Any`.

I did that, and also updated the tests.
 
> I think it's not accepted, because I guess mypy or pyright will complain about this.

mypy did complain so I had to only feed it `Option`s or `str`s at a time. Fixed the `option.py` example and removed dict example.

> I think we should add a type checker to the examples on the GHA.

I am a bit confused, is it something different than what mypy checks already do?

aisk commented on src/pick/__init__.py:
```diff
@@ -129,9 +148,16 @@ def draw(self, screen: "curses._CursesWindow") -> None:
 
         for line in lines_to_draw:
-            screen.addnstr(y, x, line, max_x - 2)
+            screen.addnstr(y, x, line, max_x // 2 - 2)
```
If there is no descriptions at all, should we draw the line with half of the screen?

aisk replied:
I think we should not care too much about the performance here; iterating a list with at most hundreds of elements does not take too much resources. I think we can just caculate the `description_present` in this function.

iamalisalehi replied:
Is [87540ac](https://github.com/wong2/pick/pull/114/commits/87540acf464881cb6e06bdf81a24221fdcb7503c) good?

iamalisalehi commented:
Thank you for the great reviews. I learned a lot from you ;-)
